### ISSUE TITLE
Sass values do not work as Map keys, do not support equality checks.

### ISSUE THREAD

**Issue by chriseppstein** [sass/node-sass#911]
```js
var sass = require("node-sass")
var map = new Map();
map.set(sass.types.String("foo"), sass.types.String("bar"))
map.forEach(function(value, key){ console.log(key.getValue() + ": " + value.getValue()) });
// outputs: foo: bar
map.get(sass.types.String("foo"))
// => undefined
```

It only works if you use the **same instance** to get the value. which is not how maps are supposed to work.

It probably has something to do with this:

```js
sass.types.String("foo") === sass.types.String("foo")
// => false
```

**Comment by chriseppstein** [comment]
IMO, this should block 3.0 unless the fix is very destabilizing.

Ping @matryo

**Comment by mgreter** [comment]
AFAIR I saw the PR when Sass_Values got introduces to node-sass, and I think to remember that the equal operator was not overloaded. I did it for some obvious ones in `[perl-libsass]` [1]. Not sure how or if you can overload operators in node.js. If the objects always gets coerced to a string, it could work by overloading the toString method on the object (which I know is doable in plain javascript).

@chriseppstein can you give us a list of all operations that you would expect to work. IMO it would be easier to implement them all in one go and I'm sure I miss a lot of them in `perl-libsass` too!

[1]: https://github.com/sass/perl-libsass/blob/develop/lib/CSS/Sass/Type.pm#L42-L46

**Comment by chriseppstein** [comment]
@mgreter I spent a while reading the NAN api and couldn't find *any* documentation describing how equality works.

It may be because we've registered functions instead of properties. It would be easier and more idiomatic to interact with `color.r` instead of `color.getR()`. This causes all sass value objects to console.log as `{}`.

I'd love to be able to give you a list, but I'm just not familiar enough with js internals to do so. I'm just using things and bumping up against failed expectations... I'll keep letting you know when I find them.

I'll see if I can figure out how to expose some properties and see if this makes the objects more well behaved.

**Comment by saper** [comment]
I tend to think that what we have are just very simple wrappers and I wouldn't really expect to become a full blown equivalents of JavaScript string or whatever....

**Comment by chriseppstein** [comment]
@saper ಠ_ಠ

**Comment by chriseppstein** [comment]
Ok, I was able to convert Number to use properties instead of methods. I find this API much easier to work with as javascript user and the c++ code is actually simpler. Changing to properties would definitely break some consumer code, but it still before release... so I'm not sure what policy you guys have on that. It would be possible to expose both, but have the methods be deprecated from the git-go.
 
https://gist.github.com/chriseppstein/3a2658d12cc44536b79a

Sadly, this did not fix the equality problem :angry: I'm sure there is a way. I filed a bug against the nan library to see if they have any advice. (their project name makes googling for an answer basically impossible.) https://github.com/iojs/nan/issues/325

**Comment by saper** [comment]
Until we really produce JavaScript native objects the following works:

````
var sass = require("node-sass")
var map = new Map();
map.set(sass.types.String("foo").getValue(), sass.types.String("bar").getValue())
map.forEach(function(value, key){ console.log(key+ ": " + value) });
// outputs: foo: bar
console.log("Two")
console.log(map.get(sass.types.String("foo").getValue()))
// outputs: bar
````

https://gist.github.com/aa0a959939f64925c0ef

I know it's crude and not nice but I'd advice to unbox Sass value types into something real for any serious processing.

**Comment by chriseppstein** [comment]
ok. so I did more testing and maps don't work with any objects whether they are from node sass or native. Silly me for trusting the spec. Also, equality is only an identity check for objects so that was also my mistake.

That said, after investigating there are two things I think we can do better here. The first is that we can switch to properties. They are a better interface that the get/set methods. Second, I'd like to find a way to get these object to console.log correctly. I think this requires that we write a property enumerator. If I can make these work, would this patch be accepted?

**Comment by saper** [comment]
Just had a quick glimpse at the v8 code and it does not really need equality checks as it internalizes (as in Lisp symbols) some strings, especially those used in the hash tables as keys. So it ends up comparing pointers or hashes that are attached to objects.

I am checking out how difficult it would be to just provide primitive types and arrays. Maybe we could have a [100, 'px'] for numbers and [r, g, b, a] for colors? Would that work for you?

I am using [Jython](http://www.jython.org/) a lot to do Java prototyping and it goes very far to bring Python stuff instead of Java objects, it's pretty amazing.

**Comment by chriseppstein** [comment]
> Would that work for you?

No. I think this would be a step in the wrong direction. For an encoding like this, I'd want it to be done external to the node-sass API. The current types are good. I just think they need some small tweaks.

**Comment by nschonni** [comment]
@chriseppstein you might want to check out http://adripofjavascript.com/blog/drips/object-equality-in-javascript.html as it probably can't use strict equality for `sass.types.String("foo") === sass.types.String("foo")`
You can test using `==`, or looking at the implementations of `isEqual` methods in [Underscore](http://underscorejs.org/#isEqual) and [Lo-Dash](http://lodash.com/docs#isEqual)

**Comment by xzyfer** [comment]
Apologies, just catching up here. I'm understanding correct this is where I think we're at (please correct me if I'm wrong).

>Sass values do not work as Map keys

It's been deduced that this is a JavaScript spec issue and not a node-sass one?

>Sass values do not support equality checks

I agree this is certainly not ideal. The current Sass values types are 'boolean`, `color`, `list`, `map`, `null`, `number`, `string` all of which map neatly to native JavaScript equivalent.

I imagine it would be ideal, when in JavaScript, for the Sass value types to be act like their native JavaScript counter parts?

 >Changing [getter/setter functions] to properties

I agree with @chriseppstein here. We may be using C under the covers but that's implementation detail that we should, to the best of our abilities, not detract from the JavaScript interface. Object properties are _The JavaScript Way_ and it's an idiom we need to respect.

**Comment by xzyfer** [comment]
If I've understood correct, we should create new issues to track the two items.

Libsass 3.2.0 is a significant release that users are eagerly waiting for, and this is the final obstacle to 3.0.0.

The custom importer, and custom function APIs are marked as _experimental, use with caution_ so we have some freedom with regards to breaking changes. 

If it looks like these issues are going take substantial time to implement my preference would be to get 3.0.0 out the door.

**Comment by saper** [comment]
Yes, it's not something we can quickly change for 3.0 :(

**Comment by saper** [comment]
I don't think it's a "bug", it's an undesirable feature :) I am not sure we have type for `color`, also `number` will be problematic because of units I think. Supplying arrays for those will be difficult because we don't know if [2,3,4,4]  should be interpreted as a color or a list. But maybe we can propose some classes in JavaScript and use them in a binding, with properties as @chriseppstein suggests.

You can't even add `Numbers`:

````
> var dwa = sass.types.Number(2)
undefined
> var trzy = sass.types.Number(3)
undefined
> dwa
{}
> trzy
{}
> dwa + trzy
'[object SassNumber][object SassNumber]'
````

**Comment by mgreter** [comment]
@saper as a side note I somewhat suffer from the same problems in perl-libsass, but there I have the opportunity for operator overloading. But libsass is i.e. missing functionality to have third party bindings to access the sass_value operations, so ATM each binding would need to implement ie. string concatenation (or equality checks) on its own. I currently plan to refactor the sass_value interface to create a solid base to be used by libsass and also by external bindings. But I cannot give any ETA on this. IMO node-sass can only really implement this with custom native objects (which I think will give more flexibility for custom operations).

**Comment by saper** [comment]
Well I think we can (haven't been able to try yet) to provide native JavaScript classes and leave all Sass Value stuff to be just internal implementation. Can you outline your idea for the new sass_value stuff? (Just before myself or somebody gets around to actually implement more native types).

**Comment by mgreter** [comment]
I'm still in the experimental phase, since I need to upgrade my c++ foo (Big Three, move semantics and more). Basically I would like to have native Sass_Value types in C++ which overload assign, subtract, division, multiplication and equal operation (maybe more). This is not really a must, but I figured it should come in handy when working in C++ code (it should take away some complexity in Sass_Value AST_Nodes implementation). Next I want to expose a generic `sass_op` function (`sass_opt(enum OP_TYPE, sass_value a, sass_value b`) which should return the a new sass_value containing the result (so bindings don't need to implement this themselves).

I already posted a [WIP branch] [1], but I have hit a wall with memory handling there. I think I know the underlying problem now and I may need another approach to correctly support it the way I want.

[1]: https://github.com/sass/libsass/pull/1161

**Comment by saper** [comment]
Ouch, looks very compilicated.

Will we still be able to use libsass in plain C after this?

My C++ foo is much worse than yours, but I'd like to help with memory handling issues; current code is already a bit too complicated and it's easy to make mistakes already. For example, I am still not sure whether custom  importer output is freed or not.
